File: olnatura-qr/web/qr-enterprise-frontend/src/utils/credentialRules.ts
/** Reglas de correo corporativo y contraseña (alta de usuario). */

export const ALLOWED_EMAIL_DOMAIN = "@olnatura.com";

const BASIC_EMAIL = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

export type PasswordCheck = {
  minLength: boolean;
  upper: boolean;
  lower: boolean;
  digit: boolean;
};

export function isAllowedEmail(email: string): boolean {
  const e = (email ?? "").trim().toLowerCase();
  if (!BASIC_EMAIL.test(e)) return false;
  return e.endsWith(ALLOWED_EMAIL_DOMAIN);
}

export function emailValidationMessage(email: string): string | null {
  const raw = (email ?? "").trim();
  if (!raw) return "El correo es obligatorio.";
  const e = raw.toLowerCase();
  if (!BASIC_EMAIL.test(e)) return "El correo no tiene un formato válido.";
  if (!e.endsWith(ALLOWED_EMAIL_DOMAIN)) {
    return "Solo se permiten correos @olnatura.com.";
  }
  return null;
}

export function passwordChecks(password: string): PasswordCheck {
  const p = password ?? "";
  return {
    minLength: p.length >= 8,
    upper: /[A-Z]/.test(p),
    lower: /[a-z]/.test(p),
    digit: /[0-9]/.test(p),
  };
}

export function isValidPassword(password: string): boolean {
  const c = passwordChecks(password);
  return c.minLength && c.upper && c.lower && c.digit;
}

export function passwordValidationMessage(password: string): string | null {
  if (!(password ?? "").length) return "La contraseña es obligatoria.";
  const c = passwordChecks(password);
  const missing: string[] = [];
  if (!c.minLength) missing.push("mínimo 8 caracteres");
  if (!c.upper) missing.push("una mayúscula");
  if (!c.lower) missing.push("una minúscula");
  if (!c.digit) missing.push("un número");
  if (missing.length === 0) return null;
  return `La contraseña debe incluir: ${missing.join(", ")}.`;
}

export const PASSWORD_RULE_LABELS: { key: keyof PasswordCheck; label: string }[] = [
  { key: "minLength", label: "Mínimo 8 caracteres" },
  { key: "upper", label: "Una letra mayúscula" },
  { key: "lower", label: "Una letra minúscula" },
  { key: "digit", label: "Un número" },
];

File: olnatura-qr/api/olnaturaqr/src/main/java/com/company/olnaturaqr/support/security/CredentialRules.java
package com.company.olnaturaqr.support.security;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Reglas de correo corporativo y contraseña para altas de usuario.
 * No aplica a OAuth/JWT/Dynamics.
 */
public final class CredentialRules {

    public static final String ALLOWED_EMAIL_DOMAIN = "@olnatura.com";

    private static final Pattern HAS_UPPER = Pattern.compile("[A-Z]");
    private static final Pattern HAS_LOWER = Pattern.compile("[a-z]");
    private static final Pattern HAS_DIGIT = Pattern.compile("[0-9]");
    private static final Pattern BASIC_EMAIL = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");

    private CredentialRules() {}

    public static boolean isAllowedEmail(String email) {
        if (email == null || email.isBlank()) return false;
        String e = email.trim().toLowerCase(Locale.ROOT);
        if (!BASIC_EMAIL.matcher(e).matches()) return false;
        return e.endsWith(ALLOWED_EMAIL_DOMAIN);
    }

    public static String emailError(String email) {
        if (email == null || email.isBlank()) {
            return "El correo es obligatorio.";
        }
        String e = email.trim().toLowerCase(Locale.ROOT);
        if (!BASIC_EMAIL.matcher(e).matches()) {
            return "El correo no tiene un formato válido.";
        }
        if (!e.endsWith(ALLOWED_EMAIL_DOMAIN)) {
            return "Solo se permiten correos " + ALLOWED_EMAIL_DOMAIN + ".";
        }
        return null;
    }

    public static boolean isValidPassword(String password) {
        return passwordIssues(password).isEmpty();
    }

    public static List<String> passwordIssues(String password) {
        List<String> issues = new ArrayList<>();
        if (password == null || password.isEmpty()) {
            issues.add("La contraseña es obligatoria.");
            return issues;
        }
        if (password.length() < 8) {
            issues.add("Mínimo 8 caracteres.");
        }
        if (!HAS_UPPER.matcher(password).find()) {
            issues.add("Al menos una letra mayúscula.");
        }
        if (!HAS_LOWER.matcher(password).find()) {
            issues.add("Al menos una letra minúscula.");
        }
        if (!HAS_DIGIT.matcher(password).find()) {
            issues.add("Al menos un número.");
        }
        return issues;
    }

    public static String passwordError(String password) {
        List<String> issues = passwordIssues(password);
        if (issues.isEmpty()) return null;
        return String.join(" ", issues);
    }
}
